File: src/main/java/spring/springserver/global/data/ErrorResponse.java
package spring.springserver.global.data;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Builder;

import java.time.LocalDateTime;
import java.util.Map;

@Builder
@JsonInclude(JsonInclude.Include.NON_NULL)
public record ErrorResponse(
    String code,
    String message,
    LocalDateTime timestamp,
    Map<String, String> details
) {

    public static ErrorResponse of(String code, String message) {

        return ErrorResponse.builder()
            .code(code)
            .message(message)
            .timestamp(LocalDateTime.now())
            .build();
    }
    
    public static ErrorResponse of(String code, String message, Map<String, String> details) {

        return ErrorResponse.builder()
            .code(code)
            .message(message)
            .timestamp(LocalDateTime.now())
            .details(details)
            .build();
    }
}

File: src/main/java/spring/springserver/global/data/ErrorResponse.kt
package spring.springserver.global.data

import com.fasterxml.jackson.annotation.JsonInclude
import java.time.LocalDateTime

@JsonInclude(JsonInclude.Include.NON_NULL)
data class ErrorResponse(
    val code: String,
    val message: String,
    val timestamp: LocalDateTime,
    val details: Map<String, String>? = null
) {
    companion object {
        fun of(code: String, message: String): ErrorResponse {
            return ErrorResponse(
                code = code,
                message = message,
                timestamp = LocalDateTime.now()
            )
        }

        fun of(
            code: String,
            message: String,
            details: Map<String, String>
        ): ErrorResponse {
            return ErrorResponse(
                code = code,
                message = message,
                timestamp = LocalDateTime.now(),
                details = details
            )
        }
    }
}
